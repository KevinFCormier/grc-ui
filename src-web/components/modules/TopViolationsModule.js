/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Button, Dropdown, DropdownItem } from 'carbon-components-react'
import resources from '../../../lib/shared/resources'
import config from '../../../lib/shared/config'
import msgs from '../../../nls/platform.properties'
import _ from 'lodash'

resources(() => {
  require('../../../scss/module-top-violations.scss')
})

const MORE_THRESHHOLD = 6

const TopViolationSelections = Object.freeze({
  clusters: 'clusters',
  policies: 'policies',
})

export default class TopViolationsModule extends React.Component {

  constructor (props) {
    super(props)
    const {viewState: {topViolationChoice=TopViolationSelections.clusters}} = props
    this.state = {
      showAll: false,
      topViolationChoice
    }
    this.onChange = this.onChange.bind(this)
  }

  render() {
    const { showAll } = this.state
    let cardData = this.getCardData()
    if (cardData.length>0) {
      const xcess = cardData.length>MORE_THRESHHOLD
      if (xcess && !showAll) {
        cardData = cardData.slice(0,MORE_THRESHHOLD)
      }
      return (
        <div className='module-top-violations'>
          {this.renderHeader()}
          {this.renderCards(cardData)}
          {xcess && this.renderButton()}
        </div>
      )
    } else {
      return (
        <div className='module-top-violations'>
          {this.renderNoViolations()}
        </div>
      )
    }
  }

  renderNoViolations() {
    const { locale } = this.context
    const title = msgs.get('overview.no.violations.title', locale)
    const detail = msgs.get('overview.no.violations.description', locale)
    return (
      <div className='no-violations'>
        <img className='no-violations-icon' src={`${config.contextPath}/graphics/emptystate.svg`} alt={title} />
        <div className='no-violations-title'>{title}</div>
        <div className='no-violations-detail'>{detail}</div>
      </div>
    )
  }

  renderHeader() {
    const { locale } = this.context
    const { topViolationChoice } = this.state
    const choices = this.getTopViolationChoices(locale)
    const title = msgs.get('overview.top.violations.title', locale)
    return (
      <div className='header-container'>
        <div className='header-title'>{title}</div>
        <Dropdown className='selection'
          value={topViolationChoice}
          ariaLabel={title}
          onChange={this.onChange} >
          {choices.map(({text, value})=> {
            return (
              <DropdownItem key={value} itemText={text} value={value} />
            )
          })}
        </Dropdown>
      </div>
    )
  }

  renderCards(cardData) {
    return (
      <div className='card-container-container' >
        {cardData.map(({name, description, count, violationType}) => {
          return (
            <div key={Math.random()}>
              <div className='card-container'>
                <div className='card-content'>
                  <div className='card-inner-content'>
                    <div className='card-violations'>
                      <div className='card-violation-count'>
                        {count}
                      </div>
                      <div className='card-violation-type'>
                        {violationType}
                      </div>
                    </div>
                    <div className='card-name-description'>
                      <div className='card-name'>
                        {name}
                      </div>
                      <div className='card-description'>
                        {description.join(', ')}
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          )
        })}
      </div>
    )
  }

  renderButton() {
    const { showAll } = this.state
    const label = msgs.get(showAll ? 'button.show.less' : 'button.show.all', this.context.locale)
    return (
      <div className='button-container'>
        <Button small kind='secondary' id={label} onClick={this.handleToggleAll}>
          { label }
        </Button>
      </div>
    )
  }

  handleToggleAll = () => {
    this.setState(prevState=>{
      return {showAll: !prevState.showAll}
    })
  }

  getCardData = () => {
    const { locale } = this.context
    const { policies } = this.props
    const { topViolationChoice } = this.state
    const dataMap = {}
    policies.map(policy=>{
      const statuses = _.get(policy, 'raw.status.status', {})
      Object.keys(statuses).forEach(key=>{
        const status = statuses[key].compliant
        if (status!=='Compliant') {
          let name, violationType, description
          switch (topViolationChoice) {
          case TopViolationSelections.policies:
            name = _.get(policy, 'metadata.name', 'unknown')
            description = key
            violationType = msgs.get('overview.top.violations.cluster', locale)
            break
          case TopViolationSelections.clusters:
            name = key
            description = _.get(policy, 'metadata.name', 'unknown')
            violationType = msgs.get('overview.top.violations.policy', locale)
            break
          }
          let data = dataMap[name]
          if (!data) {
            violationType = msgs.get('overview.top.violations', [violationType], locale)
            dataMap[name] = data = {
              name,
              description: [],
              count: 0,
              violationType
            }
          }
          data.description.push(description)
          data.count++
        }
      })
    })
    return Object.keys(dataMap).map(key=>{
      return {...dataMap[key]}
    }).sort(({count:a},{count:b})=>{
      return b-a
    })
  }

  getTopViolationChoices = (locale) => {
    if (!this.topViolationChoices) {
      this.topViolationChoices = [
        {
          value: TopViolationSelections.clusters,
          text: msgs.get('overview.top.violations.clusters', locale),
        },
        {
          value: TopViolationSelections.policies,
          text: msgs.get('overview.top.violations.policies', locale),
        },
      ]
    }
    return this.topViolationChoices
  }

  onChange = (e) => {
    const {value} = e
    this.props.updateViewState({topViolationChoice: value})
    this.setState(()=>{
      return {topViolationChoice: value}
    })
  }
}

TopViolationsModule.propTypes = {
  policies: PropTypes.array,
  updateViewState: PropTypes.func,
  viewState: PropTypes.object,
}
