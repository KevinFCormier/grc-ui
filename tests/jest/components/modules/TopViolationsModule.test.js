/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { policiesTestingDataSet1, policiesTestingDataSet2 } from './ModuleTestingData'
import TopViolationsModule from '../../../../src-web/components/modules/TopViolationsModule'

describe('TopViolationsModule clusters view', () => {
  const viewState = {viewState: {topViolationChoice:'clusters'}}
  const activeFilters = {activeFilters:{}}
  const updateViewState = jest.fn()
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet1
  it('renders as expected', () => {
    const component = renderer.create(<BrowserRouter><TopViolationsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })
})

describe('TopViolationsModule policies view', () => {
  const viewState = {viewState: {topViolationChoice:'policies'}}
  const activeFilters = {activeFilters:{}}
  const updateViewState = jest.fn()
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet2
  it('renders as expected', () => {
    const component = renderer.create(<BrowserRouter><TopViolationsModule  viewState={viewState}
      updateViewState={updateViewState}
      policies={filteredPolicies}
      activeFilters={activeFilters}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component).toMatchSnapshot()
  })
})
