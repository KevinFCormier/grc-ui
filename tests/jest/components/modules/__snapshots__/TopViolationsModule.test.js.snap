// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopViolationsModule clusters view renders as expected 1`] = `
<div
  className="module-top-violations"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Top violations
    </div>
    <div
      className="header-tab"
    >
      <nav
        aria-label="Top violations undefined"
        className="bx--tabs"
        role="navigation"
        selected={1}
      >
        <div
          aria-label="listbox"
          className="bx--tabs-trigger"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <a
            className="bx--tabs-trigger-text"
            href="#"
            onClick={[Function]}
            tabIndex={-1}
          >
            Clusters
          </a>
          <svg
            alt="show menu options"
            aria-label="show menu options"
            className={undefined}
            fill={undefined}
            fillRule="evenodd"
            height="5"
            name={undefined}
            role="img"
            style={undefined}
            viewBox="0 0 10 5"
            width="10"
          >
            <title>
              show menu options
            </title>
            <path
              d="M0 0l5 4.998L10 0z"
            />
          </svg>
        </div>
        <ul
          className="bx--tabs__nav bx--tabs__nav--hidden"
          role="tablist"
        >
          <li
            className="bx--tabs__nav-item"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={false}
            tabIndex={-1}
          >
            <a
              aria-selected={false}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={-1}
            >
              Policies
            </a>
          </li>
          <li
            className="bx--tabs__nav-item bx--tabs__nav-item--selected"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={true}
            tabIndex={-1}
          >
            <a
              aria-selected={true}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={0}
            >
              Clusters
            </a>
          </li>
        </ul>
      </nav>
      <div
        aria-hidden={true}
        className={undefined}
        hidden={true}
        selected={false}
      />
      <div
        aria-hidden={false}
        className={undefined}
        hidden={false}
        selected={true}
      />
    </div>
  </div>
  <div>
    <div
      className="violation-card-container"
    >
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  3
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  cluster1
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/policy-comp/compliance-all"
                    onClick={[Function]}
                  >
                    compliance-all
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/policy-comp/policy-ma"
                    onClick={[Function]}
                  >
                    policy-ma
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/policy-comp/policy-pod"
                    onClick={[Function]}
                  >
                    policy-pod
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  2
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  clusterhub
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/policy-comp/compliance-all"
                    onClick={[Function]}
                  >
                    compliance-all
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/policy-comp/policy-pod"
                    onClick={[Function]}
                  >
                    policy-pod
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopViolationsModule findings clusters view renders as expected 1`] = `
<div
  className="module-top-violations"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Top security findings
    </div>
    <div
      className="header-tab"
    >
      <nav
        aria-label="Top security findings undefined"
        className="bx--tabs"
        role="navigation"
        selected={0}
      >
        <div
          aria-label="listbox"
          className="bx--tabs-trigger"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <a
            className="bx--tabs-trigger-text"
            href="#"
            onClick={[Function]}
            tabIndex={-1}
          >
            Security findings
          </a>
          <svg
            alt="show menu options"
            aria-label="show menu options"
            className={undefined}
            fill={undefined}
            fillRule="evenodd"
            height="5"
            name={undefined}
            role="img"
            style={undefined}
            viewBox="0 0 10 5"
            width="10"
          >
            <title>
              show menu options
            </title>
            <path
              d="M0 0l5 4.998L10 0z"
            />
          </svg>
        </div>
        <ul
          className="bx--tabs__nav bx--tabs__nav--hidden"
          role="tablist"
        >
          <li
            className="bx--tabs__nav-item bx--tabs__nav-item--selected"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={true}
            tabIndex={-1}
          >
            <a
              aria-selected={true}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={0}
            >
              Security findings
            </a>
          </li>
          <li
            className="bx--tabs__nav-item"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={false}
            tabIndex={-1}
          >
            <a
              aria-selected={false}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={-1}
            >
              Clusters
            </a>
          </li>
        </ul>
      </nav>
      <div
        aria-hidden={false}
        className={undefined}
        hidden={false}
        selected={true}
      />
      <div
        aria-hidden={true}
        className={undefined}
        hidden={true}
        selected={false}
      />
    </div>
  </div>
  <div>
    <div
      className="violation-card-container"
    >
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  3
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  <a
                    className="card-name-link"
                    href="undefined/policies/all/Policy%20that%20is%20not%20compliant"
                    onClick={[Function]}
                  >
                    Policy that is not compliant
                  </a>
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/mycluster_name/undefined"
                    onClick={[Function]}
                  >
                    mycluster_name
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/cluster1/undefined"
                    onClick={[Function]}
                  >
                    cluster1
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/clusterhub/undefined"
                    onClick={[Function]}
                  >
                    clusterhub
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopViolationsModule findings view renders as expected 1`] = `
<div
  className="module-top-violations"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Top security findings
    </div>
    <div
      className="header-tab"
    >
      <nav
        aria-label="Top security findings undefined"
        className="bx--tabs"
        role="navigation"
        selected={0}
      >
        <div
          aria-label="listbox"
          className="bx--tabs-trigger"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <a
            className="bx--tabs-trigger-text"
            href="#"
            onClick={[Function]}
            tabIndex={-1}
          >
            Security findings
          </a>
          <svg
            alt="show menu options"
            aria-label="show menu options"
            className={undefined}
            fill={undefined}
            fillRule="evenodd"
            height="5"
            name={undefined}
            role="img"
            style={undefined}
            viewBox="0 0 10 5"
            width="10"
          >
            <title>
              show menu options
            </title>
            <path
              d="M0 0l5 4.998L10 0z"
            />
          </svg>
        </div>
        <ul
          className="bx--tabs__nav bx--tabs__nav--hidden"
          role="tablist"
        >
          <li
            className="bx--tabs__nav-item bx--tabs__nav-item--selected"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={true}
            tabIndex={-1}
          >
            <a
              aria-selected={true}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={0}
            >
              Security findings
            </a>
          </li>
          <li
            className="bx--tabs__nav-item"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={false}
            tabIndex={-1}
          >
            <a
              aria-selected={false}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={-1}
            >
              Clusters
            </a>
          </li>
        </ul>
      </nav>
      <div
        aria-hidden={false}
        className={undefined}
        hidden={false}
        selected={true}
      />
      <div
        aria-hidden={true}
        className={undefined}
        hidden={true}
        selected={false}
      />
    </div>
  </div>
  <div>
    <div
      className="violation-card-container"
    >
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  3
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  <a
                    className="card-name-link"
                    href="undefined/policies/all/Policy%20that%20is%20not%20compliant"
                    onClick={[Function]}
                  >
                    Policy that is not compliant
                  </a>
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/mycluster_name/undefined"
                    onClick={[Function]}
                  >
                    mycluster_name
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/cluster1/undefined"
                    onClick={[Function]}
                  >
                    cluster1
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/clusterhub/undefined"
                    onClick={[Function]}
                  >
                    clusterhub
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TopViolationsModule policies view renders as expected 1`] = `
<div
  className="module-top-violations"
>
  <div
    className="header-container"
  >
    <div
      className="header-title"
    >
      Top violations
    </div>
    <div
      className="header-tab"
    >
      <nav
        aria-label="Top violations undefined"
        className="bx--tabs"
        role="navigation"
        selected={1}
      >
        <div
          aria-label="listbox"
          className="bx--tabs-trigger"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="listbox"
          tabIndex={0}
        >
          <a
            className="bx--tabs-trigger-text"
            href="#"
            onClick={[Function]}
            tabIndex={-1}
          >
            Clusters
          </a>
          <svg
            alt="show menu options"
            aria-label="show menu options"
            className={undefined}
            fill={undefined}
            fillRule="evenodd"
            height="5"
            name={undefined}
            role="img"
            style={undefined}
            viewBox="0 0 10 5"
            width="10"
          >
            <title>
              show menu options
            </title>
            <path
              d="M0 0l5 4.998L10 0z"
            />
          </svg>
        </div>
        <ul
          className="bx--tabs__nav bx--tabs__nav--hidden"
          role="tablist"
        >
          <li
            className="bx--tabs__nav-item"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={false}
            tabIndex={-1}
          >
            <a
              aria-selected={false}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={-1}
            >
              Policies
            </a>
          </li>
          <li
            className="bx--tabs__nav-item bx--tabs__nav-item--selected"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="presentation"
            selected={true}
            tabIndex={-1}
          >
            <a
              aria-selected={true}
              className="bx--tabs__nav-link"
              href="#"
              role="tab"
              tabIndex={0}
            >
              Clusters
            </a>
          </li>
        </ul>
      </nav>
      <div
        aria-hidden={true}
        className={undefined}
        hidden={true}
        selected={false}
      />
      <div
        aria-hidden={false}
        className={undefined}
        hidden={false}
        selected={true}
      />
    </div>
  </div>
  <div>
    <div
      className="violation-card-container"
    >
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  3
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  cluster1
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/cluster1/1558374675051-policy-pod"
                    onClick={[Function]}
                  >
                    1558374675051-policy-pod
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/cluster1/1558382034240-policy-pod"
                    onClick={[Function]}
                  >
                    1558382034240-policy-pod
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/cluster1/policy-pod"
                    onClick={[Function]}
                  >
                    policy-pod
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className="violation-card-container"
        >
          <div
            className="violation-card-content"
          >
            <div
              className="card-inner-content"
            >
              <div
                className="card-violations"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex="0"
              >
                <div
                  className="card-violation-count alert"
                >
                  3
                </div>
              </div>
              <div
                className="card-name-description"
              >
                <div
                  className="card-name"
                >
                  clusterhub
                </div>
                <div
                  className="card-description"
                >
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/clusterhub/1558374675051-policy-pod"
                    onClick={[Function]}
                  >
                    1558374675051-policy-pod
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/clusterhub/1558382034240-policy-pod"
                    onClick={[Function]}
                  >
                    1558382034240-policy-pod
                  </a>
                  , 
                  <a
                    className="card-each-description-link"
                    href="undefined/policies/policy/clusterhub/policy-pod"
                    onClick={[Function]}
                  >
                    policy-pod
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
